pkg/endpoint/bpf.go

```diff
@@ -962,6 +1014,58 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
 	// returns changes that need to be applied to the Endpoint's bpf policy map.
 	closer, changes := e.desiredPolicy.ConsumeMapChanges()
 	defer closer()
+	changeSize := changes.Size()
+	// If the desiredPolicy will be larger than the BPF maximum after
+	// the changes we need to reset the consumed changes and either
+	// lockdown or return an inevitable error (which would happen on
+	// insertion if we proceed).
+	if e.desiredPolicy.GetPolicyMap().Len()+changeSize > policymap.MaxEntries {
+		// Revert the map changes that were just consumed.
+		e.desiredPolicy.RevertChanges(changes)
+		// We only need to go through the mechanics of
+		// lockdown once.
+		if !e.lockdown {
+			// Avoid an insertion error further down and return an
+			// error now.
+			if !option.Config.EnableEndpointLockdownOnPolicyOverflow {
+				// Do not return ErrPolicyEntryMaxExceeded as it
+				// implies that option.Config.EnableEndpointLockdownOnPolicyOverflow
+				// is "true".
+				return fmt.Errorf("policy map exceeds the map max entries limit")
+			}
+
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.EndpointID: e.ID,
+			}).Warnf("The policy map exceeds the max entries limit, %s is enabled, locking the endpoint down.",
+				option.EnableEndpointLockdownOnPolicyOverflow)
+
+			if err := e.endpointPolicyLockdown(); err != nil {
+				e.getLogger().WithFields(logrus.Fields{
+					logfields.EndpointID: e.ID,
+				}).WithError(err).Error("Failed to lockdown endpoint:" +
+					"Consider quarantining or shutting down this node.")
+				return err
+			}
+			e.lockdown = true
+		}
+		e.updatePolicyMapPressureMetric(float64(e.desiredPolicy.GetPolicyMap().Len() + changeSize))
+		return ErrPolicyEntryMaxExceeded
+	}
+
+	if e.lockdown {
+		// Coming out of lockdown, we need to make
+		// realizedPolicy and desiredPolicy the same.
+		_, _, err := e.syncPolicyMapWith(e.realizedPolicy.GetPolicyMap(), false)
+		if err != nil {
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.EndpointID: e.ID,
+			}).WithError(err).Error("Failed to come out of lockdown successfully:" +
+				"The endpoint may be compromised, consider quarantining or shutting down this node.")
+			return fmt.Errorf("Failed to come out of lockdown successfully: %w", err)
+		}
+		e.realizedPolicy = e.desiredPolicy
```

Comment:
There is other code that compares `e.realizedPolicy` and `e.desiredPolicy`; will this have an effect?

Reply:
`realizedPolicy` and `desiredPolicy` are compared to determine if `syncPolicyMap...` has been called. Their difference indicates regeneration or syncing is taking place. They are equal when their contents are the same (which they are in this case).